Snitch Cluster Offloading

Xeratec opened the pull request:
# Changelog
This PR introduces significant improvements and additions to the project, focusing on enhanced test infrastructure, driver functionality, and target support. Key updates include better separation of host and cluster code, support for mixed ISA compilation, and the addition of a new cluster target for the CONVOLVE tapeout. The changes also address critical fixes related to interrupt handling and early-execution functions, ensuring robust and reliable system behavior.

### Mixed ISA Compilation
The current approach compiles all code for both the host and cluster cores into a single library. This requires precise handling to ensure compatibility between the different instruction set architectures (ISAs) and application binary interfaces (ABIs).
This requires careful handling to avoid invalid instructions caused by mismatched ISAs between the host and cluster cores. Hence, we define four CMake variables,`ABI`, `ISA_HOST`, and `ISA_CLUSTER_SNITCH`, to specify the appropriate ISA for each core type. The ABI has to be identical to ensure correct function calls.
Furthermore, the tests are split into `src_host` and `src_cluster` directories to clearly separate code executed on the host and cluster cores. 

### cMake Build Flow
All runtime functions executed by the host core are compiled into a dedicated `runtime_host` static library and the cluster code into `runtime_cluster_<type>` (e.g. `runtime_cluster_snitch`). Additionally, the HAL layer is compiled into the `hal_host` static libary.
The final binary is seperated into two object libaries, one for the host and one for the cluster core. The host object library is linked with the `runtime_host` and `hal_host` libraries, while the cluster object library is linked with the `runtime_cluster_<type>` library. The final binary is then linked from the two object libraries.

### Devices
Additionally, this PR adds a new`devices` folder where we in the future pull in the device SDKs (such as PULP or Snitch runtime) as external libraries. The folder structure looks like this:
```
chimera-sdk
├─ drivers
│ ├─ uart
│ └─ cluster
├─ devices (NEW)
│ ├─ snitch_cluster
│ │ ├─ trampoline_snitchCluster.c
│ │ └─ snitch_runtime (external repo)
│ ├─ pulp_cluster
│ │ ├─ trampoline_pulpCluster.c
│ │ └─ pulp_runtime (external repo)
│ └─ ITA? (external repo)
├─ hal
└─ targets
```
(Some folders are omitted for simplicity)

## Warning
Special attention is required for functions that execute before the cluster core is fully initialized, such as the trampoline function and interrupt handlers. At this stage, critical resources like the stack, global pointer, and thread pointer are not yet configured. Consequently, the compiler must not generate code that allocates stack frames. To address this, such functions are implemented as naked functions, which prevent the compiler from adding prologues or epilogues that rely on stack operations.

## Added
- **Driver Functionality**:
  - Introduced a cluster driver as part of the `runtime_host` that currently includes support for the snitch cluster.
- **Device Functionality**:
  -  Introduced the `devices` folder to add device specific code and add external runtimes.
- **CONVOLVE Cluster Target**:
  - Added a new cluster target named `CONVOLVE`.
- **VSCode Configuration**
  - Add VSCode configuration to enable automatic configuration of the C/C++ extension and support for the integrated CMake build flow on the IIS workstations.
  
## Changed
- **Refactored Test Infrastructure**:
  - Split into generic and target-specific tests.
  - Organized subtests by execution environment (`host`, `snitchCluster`).
  - Clear distinction made between host code and cluster code. This allows to specify the ISA and ABI for the cluster and host code separately
- **Refactored CMake Flow**:
  - Refactored CMake build flow to collect all source into static `runtime_host` and `runtime_cluster_<type>` library and keep ISA flags private.
- **Improve README**:
  - Added explanation to use the C/C++ extension for VSCode on the IIS workstations.
  - Simplified formatting commands

## Fixed
- **ISA Compatibility and Naked Functions**:
  - Addressed invalid instructions by ensuring that the trampoline function and interrupt handlers are correctly compiled with the appropriate ISA and ABI configurations.
  - Fixed issues with improperly generated stack frames for functions executed before cluster initialization by enforcing the use of naked functions.
- **Compiler Warnings**
  - Avoid `warning: trap_vector changed binding to STB_WEAK`

viv-eth commented on driver/cluster/CMakeLists.txt:
Have you tested whether this is compatible with the OpenTitan drivers (PR #11) ?

Xeratec replied:
No, I did not. I think we agreed to first merge this PR and rebase the other one which for sure requires some changes in the CMake flow. If you want I can already do this locally to spot potential problems before merging.

viv-eth commented on targets/chimera-convolve/CMakeLists.txt:
You might want to add your name also here ;)

Xeratec replied:
Thanks, I will 😉

viv-eth commented on tests/chimera-convolve/.gitkeep:
Is this supposed to be empty?

Xeratec replied:
Yes, because I currently do not have a test for this target (I will come with the ITA tests) but I already want to provide the folder structure.

viv-eth commented on tests/chimera-host/.gitkeep:
Same as in previous comment

Xeratec replied:
See other comment

viv-eth commented on tests/chimera-open/.gitkeep:
Same as in previous comment

Xeratec replied:
See other comment

viv-eth commented on tests/generic/snitchCluster/simpleOffload/src_cluster/test_cluster.c:
Do we really want to define the `clusterInterruptHandler()` for every test?

Xeratec replied:
Good question. I assumed that we wanted to change how interrupts are handled for some applications. As far as I know, there is only one hardware interrupt signal, and the rest has to be handled by software. Thus, I would keep it for now, and if we have more data points, refactor it later if necessary.

viv-eth replied:
I understand, I can live with it for now.

viv-eth reviewed (commented):
Please check if we are compatible with #11 , other than that only minor comments. Regarding the whole offloading process, please await feedback from @Scheremo

Scheremo commented on CMakeLists.txt:
```diff
@@ -27,15 +28,50 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
 project(chimera-sdk LANGUAGES C ASM)
 
+# WIESEP: It is important to set the ISA and ABI for the host and the cluster snitch
+set(ISA_HOST rv32imc)
+set(ABI_HOST ilp32)
+set(ISA_CLUSTER_SNITCH rv32im)
+set(ABI_CLUSTER_SNITCH ilp32)
+
+message(STATUS "[CHIMERA-SDK] ISA_HOST           : ${ISA_HOST}")
+message(STATUS "[CHIMERA-SDK] ABI_HOST           : ${ABI_HOST}")
+message(STATUS "[CHIMERA-SDK] ISA_CLUSTER_SNITCH : ${ISA_CLUSTER_SNITCH}")
+message(STATUS "[CHIMERA-SDK] ABI_CLUSTER_SNITCH : ${ABI_CLUSTER_SNITCH}")
+
```
I think with the current approach there is no way to mix ABIs without everything breaking. I suggest you either remove one of the ABI variables or add an assert.

Xeratec replied:
Hmm, it works to the point where the external libraries use a common ISA subset, all code for the cluster is strictly compiled with the `ISA_HOST` and the snitch code with `ABI_CLUSTER_SNITCH`. This is also tested for a minimal setup where the host uses `RV32IMC` and the Snitch cluster `RV32IM`.

Scheremo commented on CMakeLists.txt:
```diff
@@ -27,15 +28,50 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+
+target_compile_options(runtime
+  PRIVATE
+  -O2
```
NIT: Why `O2`?

Xeratec replied:
It is an educated random choice. In my experience, `O3` often leads to problems while `O2` works well. Do you suggest something else?

Xeratec replied:
We agreed that this choice is okay and can be adapted in case we face problems in the future.

Scheremo commented on CMakeLists.txt:
```diff
@@ -27,15 +28,50 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+# WIESEP: Expose common link option
+target_link_options(runtime
+  PUBLIC
+  -nostartfiles
+  -nostdlib
```
From a non-experimental perspective `-ffreestanding` might make sense over `-nostdlib` unless we actually provide implementations for [non-freestanding](https://gcc.gnu.org/onlinedocs/gcc-6.5.0/gcc/Standards.html) headers. `-nostdlib` also includes `-nostartfiles`.

Xeratec replied:
Good point, I did not question this, but just kept it from the previous implementation. I change to use `-ffreestanding`.

Scheremo commented on README.md:
````diff
@@ -53,17 +53,43 @@ The resulting binaries will be stored in `build/bin`, and can be used within the
-python scripts/run_clang_format.py -ir tests/ --clang-format-executable=/usr/pack/riscv-1.0-kgf/pulp-llvm-0.12.0/bin/clang-format
+```
 
+## Visual Studio Code Integration
+
+To enable automatic configuration of the C/C++ extension and support for the integrated cMake build flow on the IIS workstations, add the following content to `.vscode/settings.json`:
+```json
+{
+    "cmake.configureSettings": {
+        "TOOLCHAIN_DIR": "/usr/pack/riscv-1.0-kgf/pulp-llvm-0.12.0/bin",
+        "TARGET_PLATFORM": "chimera-convolve",
+    },
+    "cmake.environment": {
+        "PATH": "/usr/pack/riscv-1.0-kgf/default/bin:${env:PATH}",
+        "LD_LIBRARY_PATH": "/usr/pack/riscv-1.0-kgf/lib64:/usr/pack/riscv-1.0-kgf/lib64",
+    }
+}
````
I'd suggest to keep this general - paths here are IIS-specific.

Xeratec replied:
I argue it is quite obvious that you have to adjust the `TOOLCHAIN_DIR`, `PATH` and `LD_LIBRARY_PATH` depending on you local setup. However, as practically speaking most users use the IIS workstation, it is nice to have a working configuration. I can add a sentence to make clear that these paths are setup specific. Also add information how to reproduce the compiler setup

Scheremo commented on drivers/cluster/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+# Add include directories from runtime to cluster_snitch
+target_include_directories(cluster_snitch
+  PRIVATE
+  $<TARGET_PROPERTY:runtime,INTERFACE_INCLUDE_DIRECTORIES>
```
Does this imply host and device share includes? This doesn't really make sense to me, could you elaborate?
It looks like you're inverting the relationship of these libraries; cluster_snitch should INTERFACE against runtime, no?

Xeratec replied:
Nice, catch. This is the part I feel the most uncomfortable about, but also spend the most time getting a working version. 
Your suggestion is what I tried initially, but the problem was that the `NUM_CLUSTER_CORES` define is part of the host headers. One option would be to share this specific header file (`soc_addr_map.h`) but this was not trivial to me as it is located in another folder and depends on the selected target (e.g. `../../targets/chimera-convolve/include/addr_maps/soc_addr_map.h`). Is there a nicer way to solve this?

Xeratec replied:
After refactoring the CMake structure this is solve in a much cleaner way be separating into `runtime_host` and `runtime_cluster_snitch`.

Scheremo commented on drivers/cluster/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+#
+# Philip Wiese <wiesep@iis.ee.ethz.ch>
+
+file(GLOB_RECURSE C_SOURCES_SNTICH
```
NIT: SNITCH, not SNTICH :)

Xeratec replied:
Thank :)

Scheremo commented on drivers/cluster/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+  $<TARGET_PROPERTY:runtime,INTERFACE_INCLUDE_DIRECTORIES>
+)
+
+target_sources(runtime INTERFACE $<TARGET_OBJECTS:cluster_snitch>)
+target_sources(runtime PRIVATE ${C_SOURCES})
```
NIT: Naming is a bit strange here, and injection into `runtime` seems non-intuitive. I'd suggest splitting this injection into two libraries `hostcode` and `devicecode` that link into `runtime` in the top-level CMakeLists to clear this up a bit.

Scheremo replied:
Playing into the comments above: I think exposing `cluster_snitch` as a target library that gets linked into the executable makes more sense. The current approach requires reproducing compile flags and makes offloading less explicit - I'm thinking if we had 5 different clusters, we'd want to be very explicit in how compile flags are handled and which cluster's offload we are using. Maybe we can discuss offline.

Xeratec replied:
This is done.

Scheremo commented on drivers/cluster/offload_snitchCluster.c:
```diff
@@ -0,0 +1,196 @@
+#include <stdint.h>
+#include <stddef.h>
+
+// Persistent trampoline function pointer for each core
+void (*_trampoline_function[NUM_CLUSTER_CORES])(void *) = {NULL};
+
+// Peristent argument storage for the trampoline function
+void *_trampoline_args[NUM_CLUSTER_CORES] = {NULL};
+
+// Persistant stack pointer storage for each core
+void *_trampoline_stack[NUM_CLUSTER_CORES] = {NULL};
```
I'd make those `static` if possible to avoid unintended tomfoolery

Scheremo replied:
FUTURE: Might also make sense (in a future PR, of course) to think about data structures for this problem - just replicating `NUM_CLUSTER_CORES` trampolines seems a bit excessive. Maybe a linked function list? function vector?

Scheremo commented on drivers/cluster/trampoline_snitchCluster.c:
```diff
@@ -0,0 +1,60 @@
+ * @warning Make sure that this function is compiled with ISA for the Snitch cores (RV32IM)
+ *
+ */
+// WIESEP: Make sure the compiler does not allocate a stack frame
+void __attribute__((naked)) _trampoline() {
+    asm volatile(
+        // Get hart ID (hardware thread ID)
+        "csrr t1, mhartid\n" // Load mhartid into a0
+
+        // Load global pointer
+        ".option push\n"
+        ".option norelax\n"          // Disable relaxation to ensure `la` behaves as expected
+        "la gp, __global_pointer$\n" // Load address of global pointer
+        ".option pop\n"
+
+        // Set thread pointer (tp) to zero
+        "mv tp, zero\n"
+
+        // Set up stack pointer
+        "la a0, _trampoline_stack\n" // Load address of _trampoline_stack
+        "slli t1, t1, 2\n"           // Multiply hart ID by 4 (size of pointer)
+        "add a0, a0, t1\n"           // Compute the address of _trampoline_stack[hartId]
+        "lw sp, 0(a0)\n"             // Load stack pointer from the computed address
+
+        // Load function pointer and arguments
+        "la a0, _trampoline_function\n" // Load address of _trampoline_function
+        "add a0, a0, t1\n"              // Compute address of _trampoline_function[hartId]
+        "lw a1, 0(a0)\n"                // Load function pointer into a1
+
+        "la a0, _trampoline_args\n" // Load address of _trampoline_args
+        "add a0, a0, t1\n"          // Compute address of _trampoline_args[hartId]
+        "lw a0, 0(a0)\n"            // Load argument pointer into a0
+
+        // Call the offloaded function
+        "jr a1\n" // Jump and link to the function pointer in a1
+    );
+}
```
If we assume we stick with the [RISC-V calling convention](https://riscv.org/wp-content/uploads/2015/01/riscv-calling.pdf), you get up to 8 integer registers for passing arguments, a0-a7, which you *could* use to directly pass function pointers, eliminating the need for globally exposed pointers.

Scheremo replied:
I think you're also missing clobbers for a0, a1, t1

Scheremo commented on tests/CMakeLists.txt:
```diff
@@ -5,6 +5,8 @@
 
-# TODO: Add target-specific test directories
+# Check if folder has a CMakeLists.txt file
+if(EXISTS ${CMAKE_CURRENT_LIST_DIR}/${TARGET_PLATFORM}/CMakeLists.txt)
+  add_subdirectory(${CMAKE_CURRENT_LIST_DIR}/${TARGET_PLATFORM})
+endif()
```
This should probably fail loudly, or at least warn - maybe add an `else` with descriptive `FATAL_ERROR` or `WARNING`?

Xeratec replied:
I would argue it should warn. It is okay, but of course not recommended, to not have tests for a target. I will add this.

Scheremo commented on tests/generic/CMakeLists.txt:
```diff
@@ -3,5 +3,10 @@
+# Philip Wiese <wiesep@iis.ee.ethz.ch>
 
-add_subdirectory(testReturnZero)
+# Add test for host core
+add_subdirectory(host)
+
+# Add test for snitchCluster
+add_subdirectory(snitchCluster)
```
I would not split `generic` into `host` and `snitchCluster`. The idea of `generic` was to contain ISA- and micro-architecture-indepenent tests (return zero, add some numbers). Since you're assuming snitch clusters, these changes could go into `chimera-open`?

Xeratec replied:
The idea here is that offloading to a snitch cluster can be done on `chimera-open` but also `chimera-convolve` and I wanted to avoid code duplication. In this regard, `generic` refers to tests that can be run with all targets. But I see that even though this is true atm, it is not guaranteed for the future. Do you see any other way that duplicating the code?

Xeratec replied:
Or include chimera-open in chimera-convolve tests

Scheremo commented on tests/generic/host/returnZero/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+# target_include_directories(${TEST_NAME}_host PUBLIC include)
+
+# WIESEP: Set the correct ISA and ABI for the host
+target_compile_options(${TEST_NAME}_host
+  PRIVATE
+  -O2
+  -march=${ISA_HOST}
+  -mabi=${ABI_HOST}
+)
```
While this isn't a bad solution, this feels a bit impractical - you're asking the end user to know about `ISA_HOST` and `ABI_HOST`, and any other compile flags necessary. I think splitting `runtime`/`chimera-sdk` into `devicecode` and `hostcode` would eliminate the need for this at the cost of having to link both into the executable. What do you think?

Xeratec replied:
I only ask them to know about the ISA and ABI flags, and I think it makes sense to be able to change the optimization level for the user code separately. Yes, this is also a very nice suggestion, happy to discuss in person about the technical details.

Xeratec replied:
After the discussion we agree to expose two libraries (host and device) with public flags and then link the host application code against the host library and device application code against device library.

Scheremo reviewed (commented):
Some comments. Functionally the changes look reasonable to me - I think we can improve implementation on the C side a bit, but that doesn't seem so urgent. CMake structure seems suboptimal to me, I think we should invest a bit of design effort into clearly disambiguating different ISAs and not push setting compile flags to the user.

Xeratec commented:
@Scheremo I implemented the feedback provided here and from our discussion.

## Changed
- Use same ABI for all objects
- Improved some printing information during CMake configuration
- Compile separately a `runtime_host` and `runtime_cluster`
- Support adding multiple test folders for one target
- Move `generic` tests `chimera-open` and `chimera-convolve` folders
- Renamed `hal` library to `hal_host`

Scheremo commented on CMakeLists.txt:
```diff
@@ -27,16 +28,83 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
 project(chimera-sdk LANGUAGES C ASM)
 
+# WIESEP: It is important to set the ISA and ABI for the host and the cluster snitch
+set(ABI ilp32)
+set(ISA_CLUSTER_SNITCH rv32im)
+set(ISA_HOST rv32imc)
```
I think we can tie this up one level lower - I'm thinking this might want to live in `targets`.

Xeratec replied:
Yes, I agree. I will change this.

Scheremo commented on CMakeLists.txt:
```diff
@@ -27,16 +28,83 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+################################################################################
+add_library(runtime_host STATIC)
+
+target_compile_options(runtime_host
+  PRIVATE
+  -O2
+)
+
+target_compile_options(runtime_host
+  PUBLIC
+  -march=${ISA_HOST}
+  -mabi=${ABI}
+)
+
+# WIESEP: Expose common link option
+target_link_options(runtime_host
+  PUBLIC
+  -march=${ISA_HOST}
+  -mabi=${ABI}
+  -nostartfiles
+  -ffreestanding
+)
```
Same comment as above:
I think we can tie this up one level lower - I'm thinking this might want to live in `targets`.

Xeratec replied:
Same, I agree. I will adapt this.

Scheremo commented on CMakeLists.txt:
```diff
@@ -27,16 +28,83 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
-target_link_libraries(chimera-sdk INTERFACE hal)
-target_link_libraries(chimera-sdk INTERFACE runtime)
-target_sources(chimera-sdk INTERFACE $<TARGET_OBJECTS:runtime>)
+################################################################################
+# Snitch Cluster Runtime Library                                               #
+################################################################################
+add_library(runtime_cluster_snitch STATIC)
+
+# WIESEP: Do not export optimization flags
+target_compile_options(runtime_cluster_snitch
+  PRIVATE
+  -O2
+)
+
+target_compile_options(runtime_cluster_snitch
+  PUBLIC
+  -march=${ISA_CLUSTER_SNITCH}
+  -mabi=${ABI}
+)
+
+# WIESEP: Expose common link option
+target_link_options(runtime_cluster_snitch
+  PUBLIC
+  -march=${ISA_CLUSTER_SNITCH}
+  -mabi=${ABI}
+  -nostartfiles
+  -ffreestanding
+)
+
```
Same comment as above:
I think we can tie this up one level lower - I'm thinking this might want to live in `targets`.

Xeratec replied:
Same, I agree. I will adapt this.

Scheremo commented on drivers/cluster/offload_snitchCluster.c:
```diff
@@ -0,0 +1,196 @@
+#include <stddef.h>
+
+// Persistent trampoline function pointer for each core
+void (*_trampoline_function[NUM_CLUSTER_CORES])(void *) = {NULL};
+
+// Peristent argument storage for the trampoline function
+void *_trampoline_args[NUM_CLUSTER_CORES] = {NULL};
+
+// Persistant stack pointer storage for each core
+void *_trampoline_stack[NUM_CLUSTER_CORES] = {NULL};
```
FUTURE: Thinking of mixed-cluster or mixed-inter-cluster-ISA chimeras, we might want to have this explicitly only be for snitch cores.

Xeratec replied:
This would indeed be a clean solution, but I am wondering how to implement this. If this is specific to the Snitch clusters, we need a data structure (either at compile time or runtime) to map the `hartid` to the device (core) type. I would leave this to another PR and live with the unnecessary storage space for the non-Snitch clusters.

Xeratec replied:
I opened a feature issue #16 for this.

Scheremo commented on drivers/cluster/trampoline_snitchCluster.c:
```diff
@@ -0,0 +1,58 @@
+ *
+ */
+// WIESEP: Make sure the compiler does not allocate a stack frame
+void __attribute__((naked)) _trampoline() {
+    asm volatile(
+        // Get hart ID (hardware thread ID)
+        "csrr t1, mhartid\n" // Load mhartid into a0
+
+        // Load global pointer
+        ".option push\n"
+        ".option norelax\n"          // Disable relaxation to ensure `la` behaves as expected
+        "la gp, __global_pointer$\n" // Load address of global pointer
+        ".option pop\n"
+
+        // Set thread pointer (tp) to zero
+        "mv tp, zero\n"
+
+        // Set up stack pointer
+        "la a0, _trampoline_stack\n" // Load address of _trampoline_stack
+        "slli t1, t1, 2\n"           // Multiply hart ID by 4 (size of pointer)
+        "add a0, a0, t1\n"           // Compute the address of _trampoline_stack[hartId]
+        "lw sp, 0(a0)\n"             // Load stack pointer from the computed address
+
+        // Load function pointer and arguments
+        "la a0, _trampoline_function\n" // Load address of _trampoline_function
+        "add a0, a0, t1\n"              // Compute address of _trampoline_function[hartId]
+        "lw a1, 0(a0)\n"                // Load function pointer into a1
+
+        "la a0, _trampoline_args\n" // Load address of _trampoline_args
+        "add a0, a0, t1\n"          // Compute address of _trampoline_args[hartId]
+        "lw a0, 0(a0)\n"            // Load argument pointer into a0
+
+        // Call the offloaded function
+        "jr a1\n" // Jump and link to the function pointer in a1
+    );
+}
```
Why not tie this up with `offload_snitchCluster.c` and make the trampolines static?

Xeratec replied:
Because the `_trampoline` is executed by the cluster, while the code in `offload_snitchCluster.c` is host code and thus compiled differently. However, one suggestion is to move this file into a new folder where we in the future pull in the device SDKs (such as PULP or Snitch runtime) as external libraries. The folder structure could look like something this:

```
chimera-sdk
├─ drivers
│ ├─ uart
│ └─ cluster
├─ devices (NEW)
│ ├─ snitch_cluster
│ │ ├─ trampoline_snitchCluster.c
│ │ └─ snitch_runtime (external repo)
│ ├─ pulp_cluster
│ │ ├─ trampoline_pulpCluster.c
│ │ └─ pulp_runtime (external repo)
│ └─ ITA? (external repo)
├─ hal
└─ targets
```
(Some folders are omitted for simplicity)

Scheremo replied:
So thematically I think this makes sense, but practically this code needs no more than rv32i. I am okay either way you skin this cat - but i think it could technically live with the same flags as either host or device.

Xeratec replied:
In this case, it would actually not work because the host uses RV32IMC but the Snitch cluster RV32IM. But I agree that technically, we could reduce the ISA from RV32IM to RV32I, but this would introduce another layer of complexity that is completely unrequited. I think it is rather straightforward that the trampoline function is compiled with the cluster ISA, as it is also executed there.

Scheremo reviewed (commented):
I think you almost perfectly nailed it this time around - remainder of my comments are on minor cmake structuring and code org. Those changes are (more or less) optional, the big picture looks very good to me!

Xeratec commented:
@Scheremo I implemented the suggested changes and added a `devices` folder with the trampoline function for the `snitch_cluster` as suggested above. Additionally, I slightly refactored the CMake structure to avoid duplication. With this it is possible to include certain tests, drivers and devices for specific targets using a mapping and a CMake macro.

For example
```
set(DEVICE_MAPPINGS
    chimera-convolve:snitch_cluster
    chimera-open:snitch_cluster
    chimera-host:
)

# Call the macro
add_chimera_subdirectories(${TARGET_PLATFORM} "Device" DEVICE_MAPPINGS)
```

Additionally, I add the libraries in the `targets`. This has the advantage that we do not create runtime libraries for devices that are not required for this target. The downside is that the `targets` subfolder has to be added before the others in the top level `CMakeLists.txt`. I added a comment to clarify this.

If you are ok with the changes, please resolve the conversation and re-approve the PR. Thanks for all the super valuable feedback!

Scheremo commented on cmake/Utils.cmake:
```diff
@@ -29,3 +29,52 @@ macro(add_target_source name)
+    if(delim_pos EQUAL -1)
+      message(WARNING "[CHIMERA-SDK] Invalid mapping entry: '${mapping}'. Skipping.")
+      continue()
+    endif()
+
+
+    # Extract key and value
+    string(SUBSTRING "${mapping}" 0 ${delim_pos} key)
+    math(EXPR value_start "${delim_pos} + 1")
+    string(SUBSTRING "${mapping}" ${value_start} -1 value)
+
+    if(key STREQUAL "${target_platform}")
+      list(APPEND included_folders ${value})
+      break()
+    endif()
+  endforeach()
+
+  string(REPLACE "," ";" included_folders "${included_folders}")
+
+  # Align output with padding
+  string(LENGTH "[CHIMERA-SDK] Enabled ${category}s" category_prefix_length)
+  math(EXPR padding_length "36 - ${category_prefix_length}")
+  if(padding_length GREATER 0)
+    string(REPEAT " " ${padding_length} padding)
+  else()
+    set(padding "")
+  endif()
+
+  # Debug: Print the folders being included
+  message(STATUS "[CHIMERA-SDK] Enabled ${category}s${padding}: ${included_folders}")
+
+  # Add subdirectories, checking for a valid CMakeLists.txt
+  foreach(folder IN LISTS included_folders)
+    if(EXISTS ${CMAKE_CURRENT_LIST_DIR}/${folder}/CMakeLists.txt)
+      add_subdirectory(${folder})
+    else()
+      message(WARNING "[CHIMERA-SDK] ${category} folder '${folder}' does not contain a valid CMakeLists.txt. Skipping.")
+    endif()
+  endforeach()
+endmacro()
```
This is clear against the declarative intent of the build system, but maybe I'm missing some insight... a switch case in the target is much easier to understand and debug, IMO, and requires no in-depth cmake knowledge.
Whether you keep this change, or revert to a simpler structure, please remove the commented version you end up not using.

Xeratec replied:
I argue that the macro makes it easier for non-experienced users to extend the SDK with new targets, as the data structure used in the end is much simpler and shorter (it is basically a dictionary in another language). This macro hides the complexity because CMake does not support dictionaries natively. But I also understand the downside and your concern. 

The problem I see is that we either have to do quite a lot of code duplication in the CMakeLists of the `drivers`, `devices` and `tests` folders to include the set of subfolders for each target or move it to a central point in the CMakeLists of the `targets` but this is not very intuitive as it is on a different folder hierarchy. Hence, I am in favor of the approach that is currently being implemented.

Ups, I forgot to delete the commented version in one place. This I can do for sure.

Xeratec replied:
@viv-eth What do you think?

Xeratec replied:
I switched from a `macro` to a` function` to ensure that the parent scope is not unintentionally modified.

Scheremo reviewed (approved):
Not a fan of the procedural macro, but I'll leave final judgement up to you. Otherwise looks good to me.
